java/compaction/compaction-core/src/main/java/sleeper/compaction/strategy/CompactionStrategyIndex.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2022-2024 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.compaction.strategy;
+
+import sleeper.core.partition.Partition;
+import sleeper.core.statestore.FileReference;
+import sleeper.core.table.TableStatus;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CompactionStrategyIndex {
+    private final List<FilesInPartition> filesInLeafPartitions;
+
+    public CompactionStrategyIndex(TableStatus tableStatus, List<FileReference> allFileReferences, List<Partition> allPartitions) {
+        this.filesInLeafPartitions = allPartitions.stream()
+                .filter(Partition::isLeafPartition)
+                .map(partition -> FilesInPartition.forPartition(tableStatus, partition.getId(), allFileReferences))
+                .collect(Collectors.toList());
+    }
+
+    public List<FilesInPartition> getFilesInLeafPartitions() {
+        return filesInLeafPartitions;
+    }
+
+    public static class FilesInPartition {
+        private final List<FileReference> filesWithJobId;
+        private final List<FileReference> filesWithNoJobIdInAscendingOrder;
+        private final String partitionId;
+        private final TableStatus tableStatus;
+
+        static FilesInPartition noFiles(TableStatus tableStatus) {
+            return new FilesInPartition(tableStatus, null, List.of(), List.of());
+        }
+
+        static FilesInPartition forPartition(TableStatus tableStatus, String partitionId, List<FileReference> allFileReferences) {
+            return new FilesInPartition(tableStatus, partitionId,
+                    allFileReferences.stream()
+                            .filter(file -> partitionId.equals(file.getPartitionId()))
+                            .filter(file -> file.getJobId() != null)
+                            .collect(Collectors.toList()),
+                    allFileReferences.stream()
+                            .filter(file -> partitionId.equals(file.getPartitionId()))
```

Comment:
Part of this issue is to avoid iterating through all the file references in the table multiple times. This iterates through them twice as many times as there are partitions. Please can we just iterate through the whole list of files once, to split it up per partition?